File: Find All Anagrams in a String.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        map<char,int> m;
        vector<int> x;
        for(int i=0;i<p.length();i++)
            m[p[i]]++;
        int l=0,r;
        int cnt=p.length();
        for(int r=0;r<s.length();r++)
        {
            if(m.find(s[r])!=m.end() && m[s[r]])
            {
               m[s[r]]--;
                cnt--;
                if(cnt==0)
                    x.push_back(l);
            }
            else{
                while(l<r)
                {
                    if(m.find(s[l])!=m.end() )
                    {
                        m[s[l]]++;
                        cnt++;
                    }
                    l++;
                    if(m.find(s[r])!=m.end() && m[s[r]])
                    {
                        r--;
                        break;
                    }
                }
                if(m.find(s[l])==m.end())
                    l=r+1;
            }
           
            
        }
        return x;
        
    }
};

File: Golang Solutions/Find All Anagrams in a String.go
func findAnagrams(s string, p string) []int {
    m:=make(map[byte]int)
    x := []int{}
    for i:=0;i<len(p);i++{
        _,ok:=m[p[i]]
        if !ok{
            m[p[i]]=1
        }else{
            m[p[i]]++
        }
    }
    l,r:=0,0
    cnt:=len(p)
    for r=0;r<len(s);r++{
        a,ok:= m[s[r]]
        if ok && a>0{
            m[s[r]]--
            cnt--
            if cnt==0{
                x=append(x,l)
            }
                    
        }else{
            for ;l<r;{
                _,ok1:=m[s[l]]
                if ok1 {
                    m[s[l]]++
                    cnt++
                }
                l++
                b,ok2:=m[s[r]]
                if ok2 && b>0{
                    r--;
                    break;
                }
            }
            _,ok3:=m[s[l]]
            if !ok3{
                l=r+1
            }    
        }
    }
    return x
}